dbms/src/Storages/DeltaMerge/Segment.cpp

```diff
@@ -373,6 +375,77 @@ bool Segment::write(DMContext & dm_context, const RowKeyRange & delete_range)
     return delta->appendDeleteRange(dm_context, delete_range);
 }
 
+bool Segment::isDefinitelyEmpty(DMContext & dm_context, const SegmentSnapshotPtr & segment_snap) const
+{
+    RUNTIME_CHECK(segment_snap->isForUpdate());
+
+    // Fast path: all packs has been filtered away
+    if (segment_snap->getRows() == 0)
+    {
+        ProfileEvents::increment(ProfileEvents::DMSegmentIsEmptyFastPath);
+        return true;
+    }
+
+    ProfileEvents::increment(ProfileEvents::DMSegmentIsEmptySlowPath);
+
+    // Build read stream
+    auto columns_to_read = std::make_shared<ColumnDefines>();
+    columns_to_read->push_back(getExtraHandleColumnDefine(is_common_handle));
+    auto read_ranges = RowKeyRanges{rowkey_range};
```

Comment:
what about using `getInputStreamModeFast`?

Reply:
I'm not sure. My intuition was to not rely on fast mode, as fast mode may change behavior at any time. Currently fast mode will additionally do filter for del_mark, while we don't need it as well.

Reply:
OK